File: src/main/java/cube8540/oauth/authentication/credentials/oauth/security/endpoint/DefaultOAuth2AccessTokenIntrospectionConverter.java
package cube8540.oauth.authentication.credentials.oauth.security.endpoint;

import cube8540.oauth.authentication.AuthenticationApplication;
import cube8540.oauth.authentication.credentials.oauth.OAuth2Utils;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenDetails;

import java.util.HashMap;
import java.util.Map;

public class DefaultOAuth2AccessTokenIntrospectionConverter implements OAuth2AccessTokenIntrospectionConverter {
    @Override
    public Map<String, Object> convertAccessToken(OAuth2AccessTokenDetails accessToken) {
        Map<String, Object> result = new HashMap<>();

        result.put(OAuth2Utils.AccessTokenIntrospectionKey.ACTIVE, true);
        result.put(OAuth2Utils.AccessTokenIntrospectionKey.CLIENT_ID, accessToken.getClientId());
        if (accessToken.getUsername() != null) {
            result.put(OAuth2Utils.AccessTokenIntrospectionKey.USERNAME, accessToken.getUsername());
        } else {
            result.put(OAuth2Utils.AccessTokenIntrospectionKey.USERNAME, null);
        }

        long expiration = accessToken.getExpiration().atZone(AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId()).toEpochSecond();
        result.put(OAuth2Utils.AccessTokenIntrospectionKey.EXPIRATION, expiration);

        String scope = String.join(" ", accessToken.getScopes());
        result.put(OAuth2Utils.AccessTokenIntrospectionKey.SCOPE, scope);
        return result;
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/oauth/security/endpoint/DefaultOAuth2AccessTokenIntrospectionConverter.kt
package cube8540.oauth.authentication.oauth.security.endpoint

import cube8540.oauth.authentication.oauth.AccessTokenIntrospectionKey
import cube8540.oauth.authentication.oauth.security.OAuth2AccessTokenDetails

class DefaultOAuth2AccessTokenIntrospectionConverter: OAuth2AccessTokenIntrospectionConverter {

    override fun convertAccessToken(accessToken: OAuth2AccessTokenDetails): Map<String, Any?> {
        val result: MutableMap<String, Any?> = HashMap()

        result[AccessTokenIntrospectionKey.ACTIVE] = !accessToken.expired
        result[AccessTokenIntrospectionKey.SCOPE] = accessToken.scopes?.joinToString(" ")

        return result
    }
}